src/coreclr/jit/lclmorph.cpp

```diff
@@ -2426,3 +2426,474 @@ void Compiler::fgExposeLocalsInBitVec(BitVec_ValArg_T bitVec)
+    unsigned    newSimdSize        = op->GetSimdSize();
+
+    assert((newSimdBaseJitType != CORINFO_TYPE_UNDEF));
+    assert((simdBaseJitType == CORINFO_TYPE_UNDEF) ||
+           ((simdBaseJitType == newSimdBaseJitType) && (simdSize == newSimdSize)));
+
+    simdBaseJitType = newSimdBaseJitType;
+    simdSize        = newSimdSize;
+}
+
+//-----------------------------------------------------------------------------
+// LclMasksCheckLclVisitor: Find the user of a lcl var and check if it is a convert to mask
+//
+class LclMasksCheckLclVisitor final : public GenTreeVisitor<LclMasksCheckLclVisitor>
+{
+public:
+    enum
+    {
+        DoPostOrder       = true,
+        UseExecutionOrder = true
+    };
+
+    LclMasksCheckLclVisitor(Compiler* compiler, GenTreeLclVarCommon* lclOp)
+        : GenTreeVisitor<LclMasksCheckLclVisitor>(compiler)
+        , lclOp(lclOp)
+    {
+    }
+
+    Compiler::fgWalkResult PostOrderVisit(GenTree** use, GenTree* user)
+    {
+        if ((*use) == lclOp)
+        {
+            switch (lclOp->OperGet())
+            {
+                case GT_STORE_LCL_VAR:
+                    // Look for:
+                    //      STORE_LCL_VAR(ConvertMaskToVector(x))
+
+                    if (lclOp->Data()->OperIsConvertMaskToVector())
+                    {
+                        convertOp = lclOp->Data()->AsHWIntrinsic();
+                    }
+                    break;
+
+                case GT_LCL_VAR:
+                    // Look for:
+                    //      ConvertVectorToMask(LCL_VAR(x)))
+
+                    if (user->OperIsConvertVectorToMask())
+                    {
+                        convertOp = user->AsHWIntrinsic();
+                    }
+                    break;
+
+                default:
+                    break;
+            }
+            return fgWalkResult::WALK_ABORT;
+        }
+        return fgWalkResult::WALK_CONTINUE;
+    }
+
+    GenTreeHWIntrinsic* convertOp = nullptr;
+
+private:
+    GenTreeLclVarCommon* lclOp;
+};
+
+//-----------------------------------------------------------------------------
+// fgLclMasksCheckLclVar: For the given lcl var, update the weights in the table.
+//
+// Arguments:
+//     lclVar - The local variable.
+//     stmt - The statement the local variable is contained in.
+//     block - The block the local variable is contained in.
+//     weightsTable - table to update.
+//
+// Returns:
+//     True if a converted local store was found.
+//
+bool Compiler::fgLclMasksCheckLcl(GenTreeLclVarCommon* lclOp,
+                                  Statement* const     stmt,
+                                  BasicBlock* const    block,
+                                  LclMasksWeightTable* weightsTable)
+{
+    // Only these can have conversions.
+    if (!lclOp->OperIs(GT_STORE_LCL_VAR) && !lclOp->OperIs(GT_LCL_VAR))
+    {
+        return false;
+    }
+
+    bool isStore = lclOp->OperIs(GT_STORE_LCL_VAR);
```

Comment:
This needs to check whether the local is address exposed. The transformation cannot be done for address exposed locals.

Reply:
> The transformation cannot be done for address exposed locals.

What's the easiest way of getting a `LCL_ADDR` from C# code? I see the use of it for stack usage, but they end up in statements which don't have any `LocalsTreeList()`.

Using `Sve.LoadVector()` I can't make a `LCL_ADDR` appear.

Reply:
It's a property of the local, not of IR, look for `LclVarDsc::IsAddressExposed`. A test case would be something that takes the address of a vector-based mask and uses the local through the address, while also using it directly -- your current transformation can end up changing it and then the operations using it through an address will break.

Reply:
There is a problem with `LCL_ADDR`....

```
            Vector<double> mask1 = Sve.CreateFalseMaskDouble(); // Create lcl mask
            Vector<double> vec1  = Vector.Create<double>(1.3);
            ConsumrAddressExposed(ref mask1); // Use as ref
            Vector<double> vec2  = Sve.Compact(mask1, vec1); // Use as mask
            Consume(vec2);
```

Creates:
```
***** BB15 [0006]
STMT00003 ( 0x01C[E-] ... 0x02A )
               [000011] --C-G------                         *  CALL      void   ChangeMatchUse:ConsumrAddressExposed[double](byref)
               [000010] ----------- arg0                    \--*  LCL_ADDR  byref  V00 loc0         [+0]
```

The problem is there are no entires in `LocalsTreeList()` for `STMT00003`. So the pass ends up skipping this statement entirely. Which means the `LCL_ADDR` is never spotted and `mask1` ends up getting converted for all the other uses/defs of it.

Only solution I can see is to drop the `LocalsTreeList()` use and instead check every statement. Thoughts on anything else we could do?

Reply:
The pass should just skip the transformation for any local that is address exposed (`LclVarDsc::IsAddressExposed()`).